rars/tools/TimerTool.java

```diff
@@ -0,0 +1,342 @@
+    private void startTimeCmpDaemon() {
+        if (timeCmp == null) {
+            timeCmp = new TimeCmpDaemon();
+        }
+    }
+
+    // Overwrites the empty parent method, called when the tool is closed
+    protected void performSpecialClosingDuties() {
+        stop();
+    }
+
+
+    /***************************  Timer controls  *****************************/
+
+    public void start() {
+        if (!running) {
+            // Start a timer that checks to see if a timer interupt needs to be raised
+            // every millisecond
+            timer.schedule(tick, 0, 1);
+            running = true;
+        }
+    }
+
+    public void play() {
+        updateTime = true;
+        startTime = System.currentTimeMillis();
+
+    }
+
+    public void pause() {
+        updateTime = false;
+        time = savedTime + System.currentTimeMillis() - startTime;
+        savedTime = time;
+    }
+
+    // Reset all of our counters to their default values
+    protected void reset() {
+        time = 0L;
+        savedTime = 0L;
+        startTime = System.currentTimeMillis();
+        tick.updateTimecmp = true;
+        timePanel.updateTime();
+        tick.reset();
+    }
+
+    // Shutdown the timer (note that we keep the TimeCmpDaemon running)
+    public void stop() {
+        updateTime = false;
+        timer.cancel();
+        running = false;
+        reset();
+    }
+
+
+    /*****************************  Timer Classes  *****************************/
+
+    // Watches for changes made to the timecmp MMIO
+    public class TimeCmpDaemon implements Observer {
+        public boolean postInterrupt = false;
+        public long value = 0L; // Holds the most recent value of timecmp writen to the MMIO
+
+        public TimeCmpDaemon() {
+            addAsObserver();
+        }
+
+        public void addAsObserver() {
+            try {
+                Globals.memory.addObserver(this, TIME_CMP_ADDRESS, TIME_CMP_ADDRESS+8);
+            } catch (AddressErrorException aee) {
+                System.out.println("Error while adding observer in Timer Tool");
+                System.exit(0);
+            }
+        }
+
+        public void update(Observable ressource, Object accessNotice) {
+            MemoryAccessNotice notice = (MemoryAccessNotice) accessNotice;
+            int accessType = ((AccessNotice)accessNotice).getAccessType();
+            // If is was a WRITE operation
+            if (accessType == 1) {
+                int address = notice.getAddress();
+                int value = notice.getValue();
+
+                // Check what word was changed, then update the corrisponding information
+                if (address == TIME_CMP_ADDRESS) {
+                    this.value = ((this.value >> 32) << 32) + value;
+                    postInterrupt = true; // timecmp was writen to
+                }
+                else if (address == TIME_CMP_ADDRESS+4) {
+                    this.value = (this.value & 0xFFFFFFFF) + (value << 32);
```

Comment:
`value << 32`  should be always 0. Did you test setting the upper 4 bytes?

Reply:
I totally forgot to test out times larger than 32-bits, I will get on that.